syntax = "proto3";

package securechat;

service SecureChat {
  // Stream for bidirectional communication
  rpc ChatStream(stream ChatMessage) returns (stream ChatMessage) {}

  // Register a new user to the chat
  rpc Register(RegisterRequest) returns (RegisterResponse) {}
}

message RegisterRequest {
  string username = 1;
  string public_key = 2; // Public key for E2E encryption
}

message RegisterResponse {
  bool success = 1;
  string message = 2;
  repeated UserInfo users = 3; // List of current users
}

message UserInfo {
  string username = 1;
  string public_key = 2;
}

message ChatMessage {
  string sender = 1;
  string recipient = 2; // Can be "ALL" for broadcast
  string encrypted_content = 3;
  bytes iv = 4; // Initialization vector for encryption
  int64 timestamp = 5;
}